src/parser/tptp/Tptp.g

```diff
@@ -637,8 +759,320 @@ fofBinaryNonAssoc[CVC4::parser::tptp::NonAssoc& na]
+        else
+        {
+          // check if any argument is a bultin node, e.g. "~", and create a
+          // lambda wrapper then, e.g. (\lambda x. ~ x)
+          for (unsigned i = 1; i < args.size(); ++i)
+          {
+            // create a lambda wrapper, e.g. (\lambda x. ~ x)
+            if (args[i].getKind() != kind::BUILTIN)
+            {
+              continue;
+            }
+            PARSER_STATE->mkLambdaWrapper(
+                args[i],
+                (static_cast<FunctionType>(args[0].getType()))
+                    .getArgTypes()[i - 1]);
+          }
+          for (unsigned i = 1; i < args.size(); ++i)
+          {
+            expr = MK_EXPR(kind::HO_APPLY, expr, args[i]);
+          }
+        }
+      }
+    )?
+  ;
+
+// atom
+// unary op formula
+// quantified formula
+// ite form
+// let form
+// tuple -- ignore
+// formula between parethesis
+
+// <thf_unitary_formula> -> <thf_atomic_formula> | ...
+// <thf_atomic_formula>  -> <thf_defined_atomic> | ...
+// <thf_defined_atomic>  -> <defined_constant> | <thf_conditional> | <thf_let> |
+//                          (<thf_conn_term>) | <defined_term>
+// <thf_conn_term>       -> <nonassoc_connective> | <assoc_connective> |
+//                          <infix_equality> | <thf_unary_connective>
+
+thfTupleForm[std::vector<CVC4::Expr>& args]
+@declarations {
+  Expr expr;
+}
+  : thfUnitaryFormula[expr]
+   { args.push_back(expr); }
+   ( COMMA_TOK thfUnitaryFormula[expr] { args.push_back(expr); } )+
+;
+
+thfUnitaryFormula[CVC4::Expr& expr]
+@declarations {
+  Kind kind;
+  std::vector< Expr > bv;
+  Expr expr2;
+  bool equal;
+}
+  : variable[expr]
+  | thfAtomicFormula[expr]
+  | LPAREN_TOK
+    thfLogicFormula[expr]
+    RPAREN_TOK
+  | NOT_TOK
+    { expr = EXPR_MANAGER->operatorOf(CVC4::kind::NOT); }
+    (thfUnitaryFormula[expr2] { expr = MK_EXPR(expr,expr2); })?
+  | // Quantified
+    thfQuantifier[kind]
+    LBRACK_TOK {PARSER_STATE->pushScope();}
+    thfBindVariable[expr]
+    {
+      bv.push_back(expr);
+    }
+    ( COMMA_TOK thfBindVariable[expr]
+     {
+       bv.push_back(expr);
+     }
+     )*
+    RBRACK_TOK COLON_TOK
+    thfUnitaryFormula[expr]
+    {
+      PARSER_STATE->popScope();
+      // handle lambda case, in which return type must be flattened and the
+      // auxiliary variables introduced in the proccess must be added no the
+      // variable list
+      //
+      // The argument flattenVars is needed in the case of defined functions
+      // with function return type. These have implicit arguments, for instance:
+      //    (define-fun Q ((x Int)) (-> Int Int) (lambda y (P x)))
+      // is equivalent to the command:
+      //    (define-fun Q ((x Int) (z Int)) Int (@ (lambda y (P x)) z))
+      // where @ is (higher-order) application. In this example, z is added to
+      // flattenVars.
+
+      // flatten body type
+      Type range = expr.getType();
+      std::vector<Expr> flattenVars;
+      if (range.isFunction())
```

Comment:
There is a utility that does this block of code (almost exactly I think): https://github.com/CVC4/CVC4/blob/master/src/parser/parser.h#L592.
Would it be easy just to call this function?

Reply:
That was actually my inspiration. IIRC though, that is used only for types, while here it's about creating the term. While the loop part is the same between the two pieces of code, what they actually produce is different.

Reply:
I changed the code to use the existing function. There will be the spurious computation of some type information, but that is negligible.